# Klaviyo Exercise

author: Brian Shin

email: brianhesungshin@gmail.com

date: 10/30/18

In [1]:
import pandas as pd
import numpy as np
import datetime

from datetime import datetime

In [2]:
df = pd.read_csv('data_science_screening_exercise_orders.csv', encoding='utf8')
print(df.shape)
df.head(2)

(13471, 4)


customer_id  gender                 date  value
0         1000       0  2017-01-01 00:11:31  198.5
1         1001       0  2017-01-01 00:29:56  338.0

## A) Assemble a dataframe with one row per customer and the following columns:

* customer_id
* gender
* most_recent_order_date
* order_count (number of orders placed by this customer)

#### Assumptions & Considerations: 
* I assumed that the date of the order and the payments happened together at the same time with no delay between the two.
* I also assumed that orders with 0 value orders were orders where coupons or promotions were used, thus a 0 dollar amount. Otherwise, I would have some questions about whether those orders are valid or errors.

In [3]:
# getting order count per customer_id
df_orders = df.groupby('customer_id')['value'].count().reset_index()
df_orders.head(2)

customer_id  value
0         1000      1
1         1001      1

In [4]:
# getting most recent order date per customer_id
df_recent = df.sort_values('date', ascending=False).groupby('customer_id').head(1).sort_values('customer_id').reset_index()
df_recent = df_recent[['customer_id', 'date']]
df_recent.head(2)

customer_id                 date
0         1000  2017-01-01 00:11:31
1         1001  2017-01-01 00:29:56

In [5]:
# joining order count and most recent order date to customer_id and gender
df_orders_count = pd.merge(df, df_orders, on='customer_id', how='left')
df_orders_count_recent = pd.merge(df_orders_count, df_recent, on='customer_id', how='left')
df_orders_count_recent.columns = ['customer_id', 'gender', 'date', 'value', 'order_count', 'most_recent_order_date']
df_orders_count_recent = df_orders_count_recent[['customer_id', 'gender', 'most_recent_order_date', 'order_count']]

df_orders_count_recent

customer_id  gender most_recent_order_date  order_count
0             1000       0    2017-01-01 00:11:31            1
1             1001       0    2017-01-01 00:29:56            1
2             1002       1    2017-02-19 21:35:31            3
3             1003       1    2017-04-26 02:37:20            4
4             1004       0    2017-01-01 03:11:54            1
5             1005       1    2017-12-16 01:39:27            2
6             1006       1    2017-05-09 15:27:20            3
7             1007       0    2017-01-01 15:59:50            1
8             1008       0    2017-12-17 05:47:48            3
9             1009       1    2017-01-01 19:27:17            1
10            1010       1    2017-04-28 13:31:11            2
11            1011       1    2017-01-01 22:04:54            1
12            1012       1    2017-01-01 22:08:16            1
13            1013       0    2017-01-01 23:51:58            1
14            1014       1    2017-01-02 00:28:42            1
15            1015       0    2017-09-21 19:58:04            4
16            1016       1    2017-01-02 01:24:13            1
17            1017       0    2017-01-02 02:08:52            2
18            1018       0    2017-01-23 01:34:52            2
19            1017       0    2017-01-02 02:08:52            2
20            1019       0    2017-01-02 02:17:03            1
21            1020       0    2017-04-03 02:42:42            3
22            1021       1    2017-03-14 00:14:21            2
23            1022       0    2017-01-02 03:36:59            1
24            1023       0    2017-03-04 21:15:18            4
25            1024       1    2017-01-02 05:19:48            1
26            1025       0    2017-05-11 19:57:33            3
27            1026       0    2017-01-02 14:52:06            1
28            1027       0    2017-01-02 15:49:44            1
29            1028       0    2017-01-22 22:27:15            2
...            ...     ...                    ...          ...
13441         9791       1    2017-12-31 07:29:37            1
13442         9792       1    2017-12-31 12:50:05            1
13443         9793       0    2017-12-31 14:18:10            1
13444         9794       1    2017-12-31 15:13:51            1
13445         9684       1    2017-12-31 15:32:59            2
13446         9795       1    2017-12-31 15:42:08            1
13447         8497       0    2017-12-31 16:12:49            7
13448         9796       0    2017-12-31 16:19:48            1
13449         9797       0    2017-12-31 16:36:34            1
13450         8973       1    2017-12-31 17:32:35            2
13451         9798       1    2017-12-31 17:36:37            1
13452         9799       1    2017-12-31 18:20:50            1
13453         9800       1    2017-12-31 18:33:23            1
13454         9555       1    2017-12-31 18:40:34            4
13455         9801       0    2017-12-31 18:53:32            1
13456         9802       1    2017-12-31 18:55:09            1
13457         9803       0    2017-12-31 18:57:12            1
13458         9804       0    2017-12-31 19:49:44            1
13459         9805       1    2017-12-31 20:07:35            1
13460         5989       0    2017-12-31 20:09:17            3
13461         9286       1    2017-12-31 20:58:40            2
13462         9806       1    2017-12-31 21:08:36            1
13463         9807       0    2017-12-31 21:24:08            1
13464         9808       1    2017-12-31 21:39:27            1
13465         9809       1    2017-12-31 21:39:33            1
13466         9810       1    2017-12-31 22:16:20            1
13467         9811       1    2017-12-31 22:24:06            1
13468         9812       0    2017-12-31 23:01:10            1
13469         9813       0    2017-12-31 23:11:02            1
13470         8658       0    2017-12-31 23:35:09            2

[13471 rows x 4 columns]

## B) Plot the count of orders per week.

#### Assumptions & Considerations: 

* I assumed that 2017-01-01 is included in week 1, as according to the actual calendar, it was technically part of week 52 of 2016. I adjusted them accordingly.

In [6]:
# to ignore warning message of future deprecation
pd.options.mode.chained_assignment = None

# turn dates into datetime data type
df_dates = df[['customer_id', 'date']]
df_dates['date'] = pd.to_datetime(df_dates['date'])
df_dates.head(2)

customer_id                date
0         1000 2017-01-01 00:11:31
1         1001 2017-01-01 00:29:56

In [7]:
# create column for datetime week number
df_dates['week_number'] = df_dates['date'].dt.week
# change 2017-01-01 to week 1, not 52 to take assumptions into consideration
df_dates['week_number'][:14] = df_dates['week_number'][:14].apply(lambda x: int(str(x).replace('52','1')))
df_dates.head(2)

customer_id                date  week_number
0         1000 2017-01-01 00:11:31            1
1         1001 2017-01-01 00:29:56            1

In [8]:
# get counts of orders per week_number
df_week_count = df_dates.groupby('week_number').count().reset_index()
df_week_count = df_week_count[['week_number', 'date']]
df_week_count.columns = ['week_number', 'week_orders_count']
df_week_count

week_number  week_orders_count
0             1                194
1             2                264
2             3                405
3             4                233
4             5                211
5             6                195
6             7                248
7             8                315
8             9                260
9            10                318
10           11                269
11           12                240
12           13                237
13           14                350
14           15                284
15           16                220
16           17                223
17           18                257
18           19                337
19           20                627
20           21                301
21           22                215
22           23                219
23           24                310
24           25                276
25           26                380
26           27                286
27           28                226
28           29                244
29           30                188
30           31                183
31           32                192
32           33                178
33           34                302
34           35                200
35           36                218
36           37                237
37           38                172
38           39                201
39           40                347
40           41                345
41           42                182
42           43                154
43           44                159
44           45                170
45           46                183
46           47                391
47           48                343
48           49                236
49           50                294
50           51                208
51           52                244

## C) Compute the mean order value for gender 0 and for gender 1.

In [9]:
# group on gender and get mean of order values
df_gender_mean = df.groupby('gender')['value'].mean()
print(df_gender_mean)
print('mean difference:', abs(df_gender_mean[0] - df_gender_mean[1]))


gender
0    363.890006
1    350.708361
Name: value, dtype: float64
mean difference: 13.1816447738


###  Do I think the difference is significant? ###

There is a mean difference of 13.18 between gender 0 and 1. Looking at the count of orders for each gender multiplied by their respective means, the 13.18 difference becomes a total gross difference of 105578.03(6759 x 363.89 - 6712 x 350.71), which is 2.19% of the total gross sales for the year (4813487.07). In the context of operations throughout a whole year, I would have to say __yes__, I think the volume of sales amplifies this 13.18 dollar mean difference into a more significant figure, being over 100k in annual revenue. Especially as business grows the following year, this difference in order amounts based on gender could be leveraged into an opportunity to maximize potential profit.

In [10]:
df_gender_sum = df.groupby('gender')['value'].sum()
print(df_gender_sum)
print('sum difference:', abs(df_gender_sum[0] - df_gender_sum[1]))
print('total sales:', abs(df_gender_sum[0] + df_gender_sum[1]))

gender
0    2459532.55
1    2353954.52
Name: value, dtype: float64
sum difference: 105578.03
total sales: 4813487.07


In [11]:
df_gender_count = df.groupby('gender')['value'].count()
print(df_gender_count)
print('total customers:', abs(df_gender_count[0] + df_gender_count[1]))


gender
0    6759
1    6712
Name: value, dtype: int64
total customers: 13471


In [12]:
df['value'].mean()

357.3221787543603